### PR TITLE
Move the adaptive sampler to a shared location.

### PR DESCRIPTION
**Pull request by jbachorik**
The rate limiting sampler is not specific to exception profiling only so it should live in a more generic space.

The change is rather harmless - just moving the actual sampler implementation to `internal-api` project. Since the `agent-bootstrap` already depends on `internal-api` no further changes are needed.

### CODE REVIEW

File: internal-api/internal-api.gradle [hunk 1]
```diff
@@ -1,3 +1,7 @@
+ext {
+  enableJunitPlatform = true
```
Comment (richardstartin) [inline internal-api/internal-api.gradle]: ✌🏻

File: internal-api/internal-api.gradle [hunk 2]
```diff
@@ -40,11 +44,21 @@ compileJava8Java {
   targetCompatibility = 8
 }
 
+compileTestJava {
+  sourceCompatibility = 8
+  targetCompatibility = 8
+}
```
Comment (richardstartin) [inline internal-api/internal-api.gradle]: Does this mean we no longer test `internal-api` on JDK7?

**Review by richardstartin — CHANGES_REQUESTED** [review]
We can't stop compiling and testing internal-api on JDK7. Perhaps extract an interface and move the implementation into a JDK8 module under utils, with a stub implementation and a factory, then load the JDK8 implementation based on a version check.

File: internal-api/src/main/java8/datadog/trace/api/sampling/AdaptiveSampler.java [hunk 3]
```diff
@@ -86,7 +86,7 @@ boolean addSample(final long limit) {
    * @param lookback the number of windows to consider in averaging the sampling rate
    * @param taskScheduler agent task scheduler to use for periodic rolls
    */
-  StreamingSampler(
+  public AdaptiveSampler(
       final Duration windowDuration,
```
Comment (richardstartin) [inline internal-api/src/main/java8/datadog/trace/api/sampling/AdaptiveSampler.java:90]: (Stating the obvious) Java time was introduced in JDK8, so this implementation would need to be moved out of `internal-api`

**Comment by jbachorik** [comment]
Well, my intention was to have just the one source root to be compiled with Java 8 to avoid creating a new module just for this one class. Apparently, it is not that easy even with the gradle black magic :(

BTW, with move to ASM 9 to support JDK 16 bytecode level the support for Java 7 will be gone as ASM itself is compiled to level 52 (Java 8) 🤷

**Comment by jbachorik** [comment]
The `AdaptiveSampler` implementation has been moved to a separate submodule of `internal-api`.

Since it is used in an already Java 8+ code in the bootstrap I would like to avoid jumping through several facade/reflection hoops in order to use it. The appropriate checks are done at the top level when the whole profiler subsystem is enabled for Java 8+.

**Comment by richardstartin** [comment]
> BTW, with move to ASM 9 to support JDK 16 bytecode level the support for Java 7 will be gone as ASM itself is compiled to level 52 (Java 8) 🤷

Yes, I agree, maintaining support for JDK7 is holding us back, but we we have users on JDK7 and can't drop it until the product decision has been made.

**Comment by jbachorik** [comment]
>  we have users on JDK7 and can't drop it until the product decision has been made

My (slightly offtopic) point was that with ASM 9 you don't have the choice any more :(

Anyway, for this PR I pulled out the Java 8 specific code to a separate module to make it more explicit (also, it makes the build much simpler).

**Comment by richardstartin** [comment]
> My (slightly offtopic) point was that with ASM 9 you don't have the choice any more :(

I got it, and this will force the issue soon, as it presents a dichotomy between JDK7 and JDK17

File: dd-java-agent/agent-bootstrap/agent-bootstrap.gradle [hunk 4]
```diff
@@ -31,7 +32,7 @@ compileMain_java11Java.options.forkOptions.javaHome = file(System.env.JAVA_11_HO
 compileMain_java11Java.sourceCompatibility = JavaVersion.VERSION_1_8
 compileMain_java11Java.targetCompatibility = JavaVersion.VERSION_1_8
 dependencies {
-  main_java11CompileOnly project(':internal-api')
+  main_java11Compile project(':internal-api:internal-api-8')
```
Comment (richardstartin) [inline dd-java-agent/agent-bootstrap/agent-bootstrap.gradle:35]: Why `compileOnly` -> `compile`?

**Comment by richardstartin** [comment]
[This](https://app.circleci.com/pipelines/github/DataDog/dd-trace-java/6680/workflows/df90c0db-113f-421d-a3f6-5e0427d40f80/jobs/132917) looks problematic

```
java.lang.NoClassDefFoundError: datadog/trace/api/sampling/AdaptiveSampler
	at datadog.trace.bootstrap.instrumentation.exceptions.ExceptionSampler.<init>(ExceptionSampler.java:25)
	at datadog.trace.bootstrap.instrumentation.exceptions.ExceptionSampler.<init>(ExceptionSampler.java:21)
	at datadog.trace.bootstrap.instrumentation.exceptions.ExceptionProfiling.<init>(ExceptionProfiling.java:26)
	at datadog.trace.bootstrap.instrumentation.exceptions.ExceptionProfiling.<clinit>(ExceptionProfiling.java:11)
	at java.lang.Class.forName(Class.java:377)
	at AgentTestRunnerTest.classpath setup(AgentTestRunnerTest.groovy:50)
```

**Comment by jbachorik** [comment]
👋  All tests are finally passing 🎉

File: internal-api/internal-api-8/internal-api-8.gradle [hunk 5]
```diff
@@ -0,0 +1,23 @@
+ext {
+  enableJunitPlatform = true
+  minJavaVersionForTests = JavaVersion.VERSION_1_8
+}
+
+apply from: "$rootDir/gradle/java.gradle"
+apply plugin: "idea"
```
Comment (tylerbenson) [inline internal-api/internal-api-8/internal-api-8.gradle:7]: This is probably not needed.

File: internal-api/internal-api-8/src/test/java/datadog/trace/api/sampling/AdaptiveSamplerTest.java [hunk 6]
```diff
@@ -38,9 +39,8 @@
  * com.datadog.profiling.exceptions.test-iterations} system property.
  */
 @ExtendWith(MockitoExtension.class)
-@Slf4j
-class StreamingSamplerTest {
-
+class AdaptiveSamplerTest {
+  private static final Logger log = LoggerFactory.getLogger(AdaptiveSamplerTest.class);
```
Comment (tylerbenson) [inline internal-api/internal-api-8/src/test/java/datadog/trace/api/sampling/AdaptiveSamplerTest.java:43]: Use `@Slf4j` on class instead.

**Review by tylerbenson — APPROVED** [review]

**Review by richardstartin — APPROVED** [review]

**Pull request merged**